Add Google Plus geolocation

icovada opened the pull request:
**Description:**
This code scrapes the data you can get from https://aboutme.google.com, then clicking the three bars at the top left, then "Overview" (i think, my Google is not in English)

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4141

**Example entry for `configuration.yaml`:**

``` yaml
device_tracker:
  - platform: gplus
    id: myphone
    url: https://aboutme.google.com/_/ProfilesMeUi/data?ds.extension=data&f.sid=-numbers&hl=it&soc-app=variables&soc-platform=1&soc-device=1&_reqid=stuff&rt=c
    token:
      - cookies:
        'SID': 'stuff'
        'HSID': 'morestuff'
      - data:
        'f.req': 'more variables'

```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

**I need your help**
I have been able to make this work by adding my cookies directly to the code, but obviously this is not a viable solution.
The example configuration.yaml is how I had thought to make it look like, but I can't seem to pass dictionaries into the code. Would someone be so kind to help me?

**How to get your data on Firefox**
1) Visit https://aboutme.google.com
2) Open the Instruments pane (Ctrl + Shift + C)
3) Click the network tab
4) click the three bar menu on the top left
5) click the first menu
6) look for a json request that looks like `data?ds.extension`
7) Right click on it, select "Copy with cURL"
8) Paste the result on http://curl.trillworks.com/#python to get your command formatted for python requests

icovada committed:
Add Google Plus geolocation

This code tracks a Google user's position through an already opened session on Google+

aa755 committed:
autopep8/autoflake

aa755 committed:
imported user-specific (sensitive) data from config file

aa755 committed:
fixed some statically checkable runtime errors

aa755 committed:
increased refresh interval default.

This should not be too high, else google may block requests.

aa755 committed:
autopep8/autoflake

aa755 commented:
@icovada : Thanks. I submitted a PR to your branch.

icovada committed:
Merge pull request #1 from aa755/gmaps

Code from aa755

aa755 commented:
It worked for me even without those missing cookies. Sorry I forgot to mention.

icovada commented:
ah! I must say I didn't even check, I saw them missing and added them

aa755 committed:
now logging in from a dummy google account.

In this setup, it is possible to share only the location of the main account
with the dummy account. This protects the main account from damage if the cookies,
which are currently stored in plain text in home-assistant config files, are leaked.

aa755 committed:
now reporting accuracy values. better config file checking

aa755 committed:
fixed a runtype type error

aa755 committed:
autoflake/autopep8

aa755 commented:
I have made more progress in my branch:
https://github.com/aa755/home-assistant/commits/gmaps

1) I also track accuracy of the location value
2) Because I was concerned about putting cookies of my main google account in the config file, I have updated the component to support the use case for logging in as another (dummy) google account. The main account's location has to be shared with the dummy account.

I cannot submit a PR to your branch because I don't have your last 2 commits (d44afa7, 4fb36a6).
I think it would be better to not have unnecessary fields in the config file. So I don't want the changes in those commits.

icovada commented:
Amazing!. I've deleted my commits, I'm waiting for your pull request

aa755 committed:
configurable host in header, to allow the dangerous aboutme.google.com mode.

aa755 committed:
autoflake/autopep8

icovada committed:
Merge pull request #2 from aa755/gmaps

tracking accuracy, logging from a dummy account.

icovada committed:
Cosmetic fixes

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+Get location from Google Maps Geolocation API.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/device_tracker.gmaps/
```
Doesn't match the platform name.

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+""".
+
+Get location from Google Maps Geolocation API.
```
It looks more that an interaction with Google+ than the Google API.

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+import logging
+import voluptuous as vol
+
+from homeassistant.const import (CONF_TOKEN,
```
Pylint error

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+    vol.Required(CONF_HSID): vol.Coerce(str),
+    vol.Required(CONF_FREQ): vol.Coerce(str),
+    vol.Required(CONF_AT): vol.Coerce(str),
+    vol.Optional(CONF_HOST, default='plus.google.com'): vol.Coerce(str),
```
I think that it would make sense to hard-code the URL.

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+
+
+#    max_accuracy = config[CONF_ACCURACY]
+    id = config[CONF_ID]
```
This is one of the pylint errors.

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+        'Accept-Language': 'en-US,en;q=0.5',
+        'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8',
+        'Referer': 'https://' + host + '/',
```
How about using string formatting here?

fabaff commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,113 @@
+        api_request = requests.post(url, headers=headers, cookies=cookies,
+                                    data=data)
+        if api_request.ok:
+            ans = api_request.text
```
I think that it would help users if there is log message if something went wrong.

sfiorini commented:
This is the best device tracker for HA. It is probably slower in updating because it depends on how often g+ is updated. However it won't fail and it won't need any other software on the phone.
Installed in "safe" mode and tracking family of 5 perfectly.
If it can be of any use, I think I have corrected most (if not all) the issues that were reported.
See the attached file with my changes.
Thanks guys.

[gplus.py.zip](https://github.com/home-assistant/home-assistant/files/584208/gplus.py.zip)

balloob commented:
The PR no longer contains any changes …

icovada committed:
Incorporate changes from sfiorini

icovada commented:
um... ooops! Fixed

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,96 @@
+""".
```
What's with this dot ? Please remove.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+    vol.Required(CONF_HSID): cv.string,
+    vol.Required(CONF_FREQ): cv.string,
+    vol.Required(CONF_AT): cv.string,
+    vol.Optional(CONF_HEADER_HOST, default='plus.google.com'): cv.string,
```
The url won't change, so let's just keep it hard coded at googles address.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+    vol.Required(CONF_AT): cv.string,
+    vol.Optional(CONF_HEADER_HOST, default='plus.google.com'): cv.string,
+    #    vol.Optional(CONF_ACCURACY, default=100): cv.positive_int,
+    vol.Optional(CONF_INTERVAL, default=1): vol.All(cv.positive_int,
```
This seems to be not used.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+            ans = api_request.text
+            matched_lines = [line for line in ans.split(
+                '\n') if "www.google.com/maps/" in line]
+            line = matched_lines[0]
```
This will give an index error if no lines matched.

All in all this scraping tactic does not seem very robust. They will add a word and all of a sudden our latitude/longitude indexes won't match anymore.

sfiorini replied:
I have been used it for 3 days and still working good, meaning the session has expired. However, like with all Google accounts the cookies are eventually expired and the process of gathering new ones will have to be repeated. So, I kinda of agree with the robustness. What's needs to be built us an automated login process to be repeated to get new cookies when none are present or expired.

sfiorini replied:
In my previous reply I meant to write that the session has NOT expired yet after 3 days.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+
+def setup_scanner(hass, config, see):
+    """Setup Scanner."""
+    import requests
```
Requests is a core dependency, please import it at the top

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+REQUIREMENTS = []
+
+CONF_INTERVAL = 'interval'
+KEEPALIVE_INTERVAL = 1
```
Not being used.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+    host = config[CONF_HEADER_HOST]
+
+    headers = {
+        'Host': host,
```
This is something that requests will set when you fetch a url, you shouldn't set it.

aa755 replied:
Indeed, it was not needed it. I've removed it in a PR to icovada's branch.

balloob commented on homeassistant/components/device_tracker/gplus.py:
```diff
@@ -0,0 +1,115 @@
+        'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8',
+        'Referer': 'https://' + host + '/',
+        'X-Same-Domain': '1',
+        'Connection': 'keep-alive',
```
This is something that requests will set when you fetch a url, you shouldn't set it.

balloob reviewed (changes requested):
My biggest concern with this is the robustness. Anyone here tried to run this for over a day or two and it still worked ? I wouldn't be surprised if Google will block you or expire your session.

aa755 committed:
removed more unnecessary fields.

Also, the url, and the data dictionary changed for me after ~14 days.
The old one stopped working.
I had to repeat the process of obtaining those values (snooping the
network traffic when visiting the google plus page in firefox).

icovada committed:
Merge pull request #3 from aa755/gmaps

removed more unnecessary fields.

balloob commented:
I am going to close this PR. Although the idea is great, having a component that requires such high maintenance (it's a changing target) is only going to cause annoyed users.

sfiorini commented:
Hi @balloob, I agree with your decision, but that gave me an idea.
Could we have a place in the repo or maybe in another repo where we keep all those components that are "not supported" for different reasons (the "Not fully baked, you are on your own" repo).
There are many components out there that are quite useful, but maybe not fully cooked. One example is the Chamberlain MyQ garage door opener. I found this hass component somewhere and with a few changes it was fully operational. Then I connected with an Alexa skill and voila!!! I can open my garage with my voice. I wish all these NOT supported components were in a bucket so that techy people would just know where to go to look for stuff instead of spending hours looking at people that forked hass that forked hass, etc.

thoughts?

Stef :)

balloob commented:
The MyQ never made it in because the developer never finished his platform. If you have a working version, please open a PR 👍 

But you're right, we could use a place that is more free for all. I do think that this should however not be under my control, but should be started by the community itself. We could even have that playground repo be the place where you introduce new components and we can promote them to built-in once mature.

aa755 commented:
It seems that it suffices to change only 1 field (`data_at`) every 14 days. Furthermore, I figured out how to automatically compute the value of the field. 
I invite others to try the new version of the component:
https://github.com/aa755/home-assistant/blob/gmaps/homeassistant/components/device_tracker/gplus.py
I am worried that the current scraping method may be specific to my account. 

The following changes are needed in config entries:
1) remove `data_at`
2) add `home_url`, the url of the Google Plus profile page of the person whose location is being tracked.
